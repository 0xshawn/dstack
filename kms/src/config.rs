use anyhow::Result;
use rocket::figment::{
    providers::{Format, Toml},
    Figment,
};
use serde::Deserialize;

pub const CONFIG_FILENAME: &str = "kms.toml";
pub const SYSTEM_CONFIG_FILENAME: &str = "/etc/kms/kms.toml";
pub const DEFAULT_CONFIG: &str = include_str!("../kms.toml");

pub fn load_config_figment(config_file: Option<&str>) -> Figment {
    let leaf_config = match config_file {
        Some(path) => Toml::file(path),
        None => Toml::file(CONFIG_FILENAME),
    };
    Figment::from(rocket::Config::default())
        .merge(Toml::string(DEFAULT_CONFIG))
        .merge(Toml::file(SYSTEM_CONFIG_FILENAME))
        .merge(leaf_config)
}

#[derive(Debug, Clone, Deserialize)]
pub(crate) struct KmsConfig {
    pub allowed_mr: AllowedMr,
    pub root_ca_cert: String,
    pub root_ca_key: String,
    pub subject_postfix: String,
    pub cert_log_dir: String,
}

#[derive(Debug, Clone)]
pub(crate) struct AllowedMr {
    pub allow_all: bool,
    pub mrtd: Vec<[u8; 48]>,
    pub rtmr0: Vec<[u8; 48]>,
    pub rtmr1: Vec<[u8; 48]>,
    pub rtmr2: Vec<[u8; 48]>,
}

impl<'de> Deserialize<'de> for AllowedMr {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct RawAllowedMr {
            #[serde(default)]
            allow_all: bool,
            #[serde(default)]
            mrtd: Vec<String>,
            #[serde(default)]
            rtmr0: Vec<String>,
            #[serde(default)]
            rtmr1: Vec<String>,
            #[serde(default)]
            rtmr2: Vec<String>,
        }

        let raw = RawAllowedMr::deserialize(deserializer)?;

        fn parse_mrlist<'d, D: serde::Deserializer<'d>>(
            list: Vec<String>,
        ) -> Result<Vec<[u8; 48]>, D::Error> {
            list.into_iter()
                .map(|s| {
                    let bytes = hex::decode(&s).map_err(serde::de::Error::custom)?;
                    bytes
                        .try_into()
                        .map_err(|_| serde::de::Error::custom("invalid MR config"))
                })
                .collect()
        }

        Ok(AllowedMr {
            allow_all: raw.allow_all,
            mrtd: parse_mrlist::<D>(raw.mrtd)?,
            rtmr0: parse_mrlist::<D>(raw.rtmr0)?,
            rtmr1: parse_mrlist::<D>(raw.rtmr1)?,
            rtmr2: parse_mrlist::<D>(raw.rtmr2)?,
        })
    }
}
