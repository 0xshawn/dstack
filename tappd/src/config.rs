use figment::Figment;
use load_config::load_config;
use serde::Deserialize;

pub const DEFAULT_CONFIG: &str = include_str!("../tappd.toml");

pub fn load_config_figment(config_file: Option<&str>) -> Figment {
    load_config("tappd", DEFAULT_CONFIG, config_file, true)
}

#[derive(Debug, Clone, Deserialize)]
pub struct Config {
    pub app_name: String,
    pub cert_file: String,
    pub key_file: String,
    pub public_logs: bool,
    pub public_sysinfo: bool,
}
