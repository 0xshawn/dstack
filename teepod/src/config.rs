use std::{net::IpAddr, path::PathBuf, str::FromStr};

use anyhow::{bail, Context, Result};
use rocket::figment::{
    providers::{Format, Toml},
    Figment,
};
use serde::{Deserialize, Serialize};

pub const CONFIG_FILENAME: &str = "teepod.toml";
pub const SYSTEM_CONFIG_FILENAME: &str = "/etc/teepod/teepod.toml";
pub const DEFAULT_CONFIG: &str = include_str!("../teepod.toml");

pub fn load_config_figment(config_file: Option<&str>) -> Figment {
    let leaf_config = match config_file {
        Some(path) => Toml::file(path),
        None => Toml::file(CONFIG_FILENAME),
    };
    Figment::from(rocket::Config::default())
        .merge(Toml::string(DEFAULT_CONFIG))
        .merge(Toml::file(SYSTEM_CONFIG_FILENAME))
        .merge(leaf_config)
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Protocol {
    Tcp,
    Udp,
}

impl FromStr for Protocol {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "tcp" => Protocol::Tcp,
            "udp" => Protocol::Udp,
            _ => bail!("Invalid protocol: {s}"),
        })
    }
}

impl Protocol {
    pub fn as_str(&self) -> &str {
        match self {
            Protocol::Tcp => "tcp",
            Protocol::Udp => "udp",
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PortRange {
    pub protocol: Protocol,
    pub from: u16,
    pub to: u16,
}

impl PortRange {
    pub fn contains(&self, protocol: &str, port: u16) -> bool {
        self.protocol.as_str() == protocol && port >= self.from && port <= self.to
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct PortMappingConfig {
    pub enabled: bool,
    pub address: IpAddr,
    pub range: Vec<PortRange>,
}

impl PortMappingConfig {
    pub fn is_allowed(&self, protocol: &str, port: u16) -> bool {
        if !self.enabled {
            return false;
        }
        self.range.iter().any(|r| r.contains(protocol, port))
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct CvmConfig {
    pub ca_cert: PathBuf,
    pub tmp_ca_cert: PathBuf,
    pub tmp_ca_key: PathBuf,
    /// The URL of the KMS server
    pub kms_url: String,
    /// The URL of the TProxy server
    pub tproxy_url: String,
    /// The maximum disk size in GB
    pub max_disk_size: u32,
    /// The start of the CID pool that allocates CIDs to VMs
    pub cid_start: u32,
    /// The size of the CID pool that allocates CIDs to VMs
    pub cid_pool_size: u32,
    /// Port mapping configuration
    pub port_mapping: PortMappingConfig,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GatewayConfig {
    pub base_domain: String,
    pub port: u16,
    pub tappd_port: u16,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Config {
    #[serde(default)]
    pub image_path: PathBuf,
    #[serde(default)]
    pub run_path: PathBuf,
    #[serde(default)]
    pub qemu_path: PathBuf,
    /// The URL of the KMS server
    pub kms_url: String,

    /// CVM configuration
    pub cvm: CvmConfig,
    /// Gateway configuration
    pub gateway: GatewayConfig,
}

impl Config {
    pub fn extract_or_default(figment: &Figment) -> Result<Self> {
        let mut me: Self = figment.extract()?;
        {
            let home = dirs::home_dir().context("Failed to get home directory")?;
            let app_home = home.join(".teepod");
            if me.image_path == PathBuf::default() {
                me.image_path = app_home.join("image");
            }
            if me.run_path == PathBuf::default() {
                me.run_path = app_home.join("vm");
            }
            if me.qemu_path == PathBuf::default() {
                let cpu_arch = std::env::consts::ARCH;
                let qemu_path = which::which(format!("qemu-system-{}", cpu_arch))
                    .context("Failed to find qemu-system-x86_64")?;
                me.qemu_path = qemu_path;
            }
        }
        Ok(me)
    }
}
