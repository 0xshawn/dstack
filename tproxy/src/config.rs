use anyhow::{Context, Result};
use ipnet::Ipv4Net;
use rocket::figment::{
    providers::{Format, Toml},
    Figment,
};
use serde::{Deserialize, Serialize};
use std::net::Ipv4Addr;

#[derive(Debug, Clone, Deserialize)]
pub struct WgConfig {
    pub public_key: String,
    pub private_key: String,
    pub listen_port: u16,
    pub ip: Ipv4Addr,
    pub client_ip_range: Ipv4Net,
    pub interface: String,
    pub config_path: String,
    pub endpoint: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PortMap {
    pub listen_addr: Ipv4Addr,
    pub listen_port: u16,
    pub target_port: u16,
    pub label: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ProxyConfig {
    pub cert_chain: String,
    pub cert_key: String,
    pub base_domain: String,
    pub config_path: String,
    pub portmap: Vec<PortMap>,
    pub tls_passthrough: TlsPassthroughConfig,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TlsPassthroughConfig {
    pub listen_addr: Ipv4Addr,
    pub listen_ports: Vec<u16>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CertbotConfig {
    pub workdir: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Config {
    pub wg: WgConfig,
    pub proxy: ProxyConfig,
    pub certbot: CertbotConfig,
}

impl Config {
    pub fn compute(&self) -> Result<ComputedConfig> {
        let tappd_port_map = self
            .proxy
            .portmap
            .iter()
            .find(|p| p.label.as_deref() == Some("tappd"))
            .context("tappd port map not found")?
            .clone();
        Ok(ComputedConfig { tappd_port_map })
    }
}

#[derive(Debug, Clone)]
pub struct ComputedConfig {
    pub tappd_port_map: PortMap,
}

pub const CONFIG_FILENAME: &str = "tproxy.toml";
pub const SYSTEM_CONFIG_FILENAME: &str = "/etc/tproxy/tproxy.toml";
pub const DEFAULT_CONFIG: &str = include_str!("../tproxy.toml");

pub fn load_config_figment(config_file: Option<&str>) -> Figment {
    let leaf_config = match config_file {
        Some(path) => Toml::file(path),
        None => Toml::file(CONFIG_FILENAME),
    };
    Figment::from(rocket::Config::default())
        .merge(Toml::string(DEFAULT_CONFIG))
        .merge(Toml::file(SYSTEM_CONFIG_FILENAME))
        .merge(leaf_config)
}
