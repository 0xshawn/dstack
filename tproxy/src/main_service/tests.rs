use super::*;
use crate::config::{load_config_figment, Config};

fn create_test_state() -> AppState {
    let figment = load_config_figment();
    let config = figment.extract::<Config>().unwrap();
    AppState::new(config)
}

#[test]
fn test_empty_config() {
    let state = create_test_state();
    let proxy_config = state.lock().generate_proxy_config().unwrap();
    insta::assert_snapshot!(proxy_config);
    let wg_config = state.lock().generate_wg_config().unwrap();
    insta::assert_snapshot!(wg_config);
}

#[test]
fn test_config() {
    let state = create_test_state();
    let info = state
        .lock()
        .new_client_by_id("test-id-0", "test-pubkey-0")
        .unwrap();
    insta::assert_debug_snapshot!(info);
    let info1 = state
        .lock()
        .new_client_by_id("test-id-1", "test-pubkey-1")
        .unwrap();
    insta::assert_debug_snapshot!(info1);
    let proxy_config = state.lock().generate_proxy_config().unwrap();
    insta::assert_snapshot!(proxy_config);
    let wg_config = state.lock().generate_wg_config().unwrap();
    insta::assert_snapshot!(wg_config);
}
