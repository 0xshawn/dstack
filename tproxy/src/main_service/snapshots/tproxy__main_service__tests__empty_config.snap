---
source: tproxy/src/main_service/tests.rs
assertion_line: 14
expression: proxy_config
---
servers:
  - type: socket
    listen: 0.0.0.0:8443
    handler:
      type: lazytls
      certificate: /etc/rproxy/certs/cert.pem
      key: /etc/rproxy/certs/key.pem
      sni: []
  - type: socket
    listen: 0.0.0.0:8090
    handler:
      type: lazytls
      certificate: /etc/rproxy/certs/cert.pem
      key: /etc/rproxy/certs/key.pem
      sni: []
